### ISSUE TITLE
"assign [class=discord] workspace 2" stopped working in 1.6

### ISSUE THREAD

**Issue by Lahvuun** [swaywm/sway#6177]
- **Sway Version:**
  - sway version 1.6

- **Debug Log:**
  - https://gist.github.com/Lahvuun/77d87eac060d0c0b59660a4dd609ebe8

- **Configuration File:**
  - https://gist.github.com/Lahvuun/a780eab14b391bcd32b52873ec6b631b

- **Description:**
  - start sway
  - open terminal
  - run `swaymsg assign "[class=discord]" workspace 2`
  - run `discord`
  
When discord runs, it first shows a (floating?) update window that gets assigned to workspace 2 correctly, but the main discord window shows up in the focused workspace instead of the one assigned. This does not happen if I downgrade to 1.5.1 using my system's package manager.

I tried bisecting it, but when I build sway 1.5.1 and wlroots from source, discord still doesn't show up in the assigned workspace. It's the same for 1.5, while 1.4 doesn't build. I'm not sure why 1.5.1 installed with portage works correctly, but not when I build it from source. The [ebuild](https://gitweb.gentoo.org/repo/gentoo.git/tree/gui-wm/sway/sway-1.5.1-r1.ebuild) doesn't seem to do anything that could cause this.

**Comment by trylletre** [comment]
can confirm that the assign doesn't work, while
```
for_window [class="discord"] move container to workspace 2
```
works fine, even though the two are supposed to be equivalent as per the manpage

using sway 1.6 from the gentoo ebuild

**Comment by RPigott** [comment]
I have long used `assign [title="^Discord$"] 4` and that one still works.

I tried reproducing on 1.5.1, but as you say it has the same behavior as 1.6. Can you provide a debug log of a case where the assignment does work using the sway version from your package manager? I can't reproduce it.

**Comment by Lahvuun** [comment]
https://gist.github.com/Lahvuun/34d0026fcf9036141b934d884f1e708e
Here is a log of 1.5.1 installed with my package manager, with the same actions performed.

**Comment by markstos** [comment]
There's a known bug with electron-based apps like Discord that they set WM_CLASS too late for matching with "assign".  For example, see: https://community.spotify.com/t5/Desktop-Linux/WM-CLASS-window-property-on-Linux/td-p/1247042

It's possible before settng WM_CLAS happened be "fast enough" to work in your case, but the issue with not setting it fast enough is with Electron, not Sway.

**Comment by Lahvuun** [comment]
I did some "bisecting" by installing every commit with my package manager and found that [this](https://github.com/swaywm/wlroots/commit/de5347d0f2a82150a4788b63d06b5e91d5ebc27a) wlroots commit is what broke it. Makes sense it worked with my package manager but not when building from source before 1.6 because portage disables the `xcb-icccm` meson option by default.

**Comment by emersion** [comment]
So the `xcb-icccm` code breaks it?

**Comment by Lahvuun** [comment]
Yes. Installing sway 1.5.1 and wlroots 0.12.0 with my package manager with icccm enabled breaks `assign` for discord, but only for the main window. The small update popup is shown first and it is assigned to the correct workspace, but the main window shows in the currently focused workspace.

**Comment by reedrs** [comment]
I believe I'm encountering a similar problem to what @markstos described above. I installed a Firefox extension which prepends labels to each of my Firefox windows (like `[ff_video]`), and I wanted to then assign each Firefox window to a different workspace. But the extension takes quite a while to modify the titles of the windows after Firefox has started.

To quote `man 5 sway` (note how it says these are equivalent):
>       assign <criteria> [→] [workspace] [number] <workspace>
>           Assigns views matching criteria (see CRITERIA for details) to workspace. The → (U+2192) is optional and cosmetic. This command is equivalent to:
>
>               for_window <criteria> move container to workspace <workspace>

Based on that, I tried to do
```
assign [title="\[ff_video\].*"] my-firefox-workspace
```
, which had no effect. But to my surprise,
```
for_window [title="\[ff_video\].*"] move container to workspace my-firefox-workspace
```
works perfectly.

Does the documentation need to be clarified to explain that there is actually a difference between these, or is `assign` intended to actually behave just like the more verbose `for_window` syntax?

**Comment by markstos** [comment]
Here's the docs from the i3wm User Manual that explains the difference between `assign` and `for_window`:

> The difference between assign and `for_window <criteria> move to workspace` is that the former will only be executed when the application maps the window (mapping means actually displaying it on the screen) but the latter will be executed whenever a window changes its properties to something that matches the specified criteria.

Based on that reference, the behavior @reedrs experienced is correct, but the current sway docs for `assign` fail to clarify that it only triggers when a window first appears. The fix I recommend is to patch the docs for `assign` to clarify when it triggers:

https://github.com/swaywm/sway/blob/master/sway/sway.5.scd

**Comment by stutteringsteve** [comment]
To provide a temporary workaround for everyone, I decided to create individual scripts for each program that is otherwise assigned a specific workstation within my Sway configuration and placed said scripts under `$PATH`.  So, for example, I have `ff` for Firefox.  Within each script, I use `swaymsg` to determine whether an assigned workspace in question is focused, and if not, I proceed with focusing it prior to launching a desired program.  To use Firefox as an example:

```
#!/bin/sh

sway_ws=1

swaymsg -t get_workspaces \
	| jq '.[] | select(.focused==true).name' \
	| cut -d"\"" -f2 \
	| grep -E "$sway_ws\:" \
	|| \
swaymsg -t get_config \
	| grep -Eo "set \\\$ws$sway_ws [0-9]+\:[a-z]+" \
	| awk -F ' ' '{print $3}' \
	| xargs swaymsg workspace

firefox "$@"
```

Notice the `sway_ws` variable at the near top of this example.  This represents an assigned workspace, in this case, for Firefox.  In this example, this script would set focus on the 1st workspace in Sway, regardless of what it's named in your Sway configuration.

However, do keep in mind that this script would take focus away from whatever workspace you're working in.  Enjoy!

**Comment by Lahvuun** [comment]
This was fixed by 21d2fdf74c93a4d1df5dd2dc0d6bf24c611dd752
